Fix race condition when closing encoders

88hcsif opened the pull request:
I am applying the principle found [here](https://go101.org/article/channel-closing.html): **don't close a channel from the receiver side and don't close a channel if the channel has multiple concurrent senders**.

Media will now wait for the emulator to gracefully close, and then close the encoder input.
The encoder will wait for media to gracefully close its input, and then close its output and release any memory that cannot be garbage collected.

There are also other channels in the project that could be refactored to follow this principle.
However, they are a bit more complex and more importantly they seem to work, so let's not change it if it ain't broken.

This should fix issue #180.

giongto35 commented on pkg/encoder/h264encoder/encoder.go:
```diff
@@ -4,12 +4,14 @@ import (
 )
 
 const chanSize = 2
+var mu sync.Mutex
```
we might have multiple encoder, so better put lock in H264Encoder

giongto35 commented on pkg/encoder/vpx-encoder/encoder.go:
```diff
@@ -42,6 +43,7 @@ GoBytesType get_frame_buffer(vpx_codec_ctx_t *codec, vpx_codec_iter_t *iter) {
 import "C"
 
 const chanSize = 2
+var mu sync.Mutex
```
similarly, put lock in VpxEncoder struct as well

giongto35 commented:
Why you need to have mutex for controlling close order?

giongto35 commented:
The way I check is after a session is closed, all goroutine is cleared. If there are multiple sessions join and leave, the number of goroutine is still a constant.
I also tried to follow the principle. I think there is someplace I feel difficulty to strictly follow it because the system is bidirectional. Emulator is the writer for image stream, but is reader for input. WebRTC is the reader for image stream but writer for game input. Room is the middleware.
Therefore, it's a bit messy for closing order right now, so It's great if you want to refactor it : D.

88hcsif commented:
Yeah it is definitely a complex architecture which needs some extra thought and that actually might be already correct as it is.

I moved the mutexes (now a channels) in the struct so that they are not global as you suggested.
I also moved the call to close the encoder input inside the stop/release function so that it should not be possible to misuse the encoder and cause a deadlock.

The reason why I was using mutexes and now channels is to coordinate the looping go routine and the stop/release function.
We cannot release memory until we know for sure that the loop is stopped.
I changed to channels because I guess they are more common and more appropriate in golang.

giongto35 commented:
Yes, the code looks good to me now :P. Thank you for the fix
